scheduler/test/cook/test/simulator.clj

```diff
@@ -0,0 +1,599 @@
+         (.close curator-framework#)
+         (.stop zookeeper-server#)
+         (doseq [[trigger-service# trigger-chan#] trigger-chans#]
+           (log/info "Shutting down" trigger-service#)
+           (async/close! trigger-chan#))))))
+
+(defn generate-task-trace-map
+  [task]
+  (let [job (:job/_instance task)
+        resources (util/job-ent->resources job)]
+    {:job_id  (str (:job/uuid job))
+     :instance_id  (:instance/task-id task)
+     :submit_time_ms  (.getTime (:job/submit-time job))
+     :mesos_start_time_ms  (if (:instance/mesos-start-time task)
+                             (.getTime (:instance/mesos-start-time task))
+                             -1)
+     :start_time_ms  (.getTime (:instance/start-time task))
+     :end_time_ms  (.getTime (or (:instance/end-time task) (tc/to-date (t/now))))
+     :status  (:instance/status task)
+     :hostname  (:instance/hostname task)
+     :slave_id  (:instance/slave-id task)
+     :reason  (or (when (= (:instance/status task) :instance.status/failed)
+                    (:reason/string (:instance/reason task)))
+                  "")
+     :user  (:job/user job)
+     :mem  (or (:mem resources) -1)
+     :cpus  (or (:cpus resources) -1)
+     :job_name  (or (:job/name job) "")
+     :requested_run_time  (->> (:job/label job)
+                               (filter #(= (:label/key %) "JOB-RUNTIME"))
+                               first
+                               :label/value)
+     :requested_status  (->> (:job/label job)
+                             (filter #(= (:label/key %) "JOB-STATUS"))
+                             first
+                             :label/value)}))
+
+(defn dump-jobs-to-csv
+  "Given a mesos db, dump a csv with a row per task
+
+   Columns:
+   * job_id
+   * instance_id
+   * submit_time
+   * mesos_start_time
+   * start_time
+   * end_time
+   * hostname
+   * slave_id
+   * status
+   * reason
+   * user
+   * mem
+   * cpus
+   * job-name
+   * requested-run-time
+   * requested-status"
+  [task-ents file]
+  ;; Use snake case to make it easier for downstream tools to consume
+  (let [headers [:job_id :instance_id :submit_time_ms :mesos_start_time_ms :start_time_ms
+                 :end_time_ms :hostname :slave_id :status :reason :user :mem :cpus :job_name
+                 :requested_run_time :requested_status]
+        tasks (map generate-task-trace-map task-ents)]
+    (with-open [out-file (io/writer file)]
+      (csv/write-csv out-file
+                     (concat [(mapv name headers)]
+                             (map (apply juxt headers) tasks))))))
+
+(defn pull-all-task-ents
+  "Returns a seq of task entities from the db"
+  [mesos-db]
+  (->> (d/q '[:find [?i ...]
+         :where
+         [?i :instance/task-id _]]
+       mesos-db)
+       (map (partial d/entity mesos-db))))
+
+(defn submit-job
+  "Submits a job to datomic"
+  [conn job]
+  (let [job-keys [:job/uuid :job/command :job/user :job/name :job/max-retries
```

Comment:
can we list the keys in sorted order, it makes it easier to find entries that are being retrieved